amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java

```diff
@@ -1324,175 +1591,452 @@ public void testDetermineNewLeasesToCreateSplitMergeHorizon2() {
         final List<Shard> shards = constructShardListForGraphA();
-        final List<Lease> currentLeases = Arrays.asList(newLease("shardId-3"), newLease("shardId-4"),
-                newLease("shardId-5"));
+        final List<String> shardIdsOfCurrentLeases = Collections.emptyList();
+        final Map<String, ExtendedSequenceNumber> expectedShardIdCheckpointMap = new HashMap<>();
+        expectedShardIdCheckpointMap.put("shardId-0", ExtendedSequenceNumber.TRIM_HORIZON);
+        expectedShardIdCheckpointMap.put("shardId-1", ExtendedSequenceNumber.TRIM_HORIZON);
+        expectedShardIdCheckpointMap.put("shardId-2", ExtendedSequenceNumber.TRIM_HORIZON);
+        expectedShardIdCheckpointMap.put("shardId-3", ExtendedSequenceNumber.TRIM_HORIZON);
+        expectedShardIdCheckpointMap.put("shardId-4", ExtendedSequenceNumber.TRIM_HORIZON);
+        expectedShardIdCheckpointMap.put("shardId-5", ExtendedSequenceNumber.TRIM_HORIZON);
+        assertExpectedLeasesAreCreated(shards, shardIdsOfCurrentLeases, INITIAL_POSITION_TRIM_HORIZON, expectedShardIdCheckpointMap);
+    }
 
-        final Map<String, Shard> shardIdToShardMap = HierarchicalShardSyncer.constructShardIdToShardMap(shards);
-        final Map<String, Set<String>> shardIdToChildShardIdsMap = HierarchicalShardSyncer
-                .constructShardIdToChildShardIdsMap(shardIdToShardMap);
+    /**
+     * Test CheckIfDescendantAndAddNewLeasesForAncestors (initial position Latest)
+     * Shard structure (each level depicts a stream segment):
+     * 0 1 2 3 4   5- shards till epoch 102
+     * \ / \ / |   |
+     *  6   7  4   5- shards from epoch 103 - 205
+     *   \ /   |  /\
+     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
+     * Current leases: (0, 1, 4, 9, 10)
+     */
+    @Test
+    public void testDetermineNewLeasesToCreateSplitMergeHorizonA_CompleteHashRangeAcrossDifferentEpochs() {
+        final List<Shard> shards = constructShardListForGraphA();
+        final List<String> shardIdsOfCurrentLeases = Arrays.asList("shardId-0", "shardId-1", "shardId-4", "shardId-7",
+                "shardId-9", "shardId-10");
+        final Map<String, ExtendedSequenceNumber> expectedShardIdCheckpointMap = Collections.emptyMap();
+        assertExpectedLeasesAreCreated(shards, shardIdsOfCurrentLeases, INITIAL_POSITION_TRIM_HORIZON, expectedShardIdCheckpointMap);
+    }
 
-        final HierarchicalShardSyncer.LeaseSynchronizer nonEmptyLeaseTableSynchronizer =
-                new HierarchicalShardSyncer.NonEmptyLeaseTableSynchronizer(shardDetector, shardIdToShardMap, shardIdToChildShardIdsMap);
+    /*
+     * Shard structure (x-axis is epochs):
+     * 0  3   6   9
+     * \ / \ / \ /
+     *  2   5   8
+     * / \ / \ / \
+     * 1  4   7  10
+     *
+     * Current leases: (3)
+     */
+    @Test
+    public void testDetermineNewLeasesToCreateSplitMergeHorizonB_PartialHashRange() {
```

Comment:
Hmmm, for this case, I feel we need to just create only 4 for both latest and trim/ts. Else we might end up processing 0,1,2,4,5 leading to kind of our of order processing. Your thoughts?

Reply:
Yeah agreed I think this is a bug. 0 and 1 are ancestors of 3, so if we're processing 3 that means we already processed 0 and 1 so we shouldn't create new leases for them. Will fix the test and the code to only create 4 in this case.

Reply:
Sent a fix for this in the new revision. It adds a bit of complexity to the code, but I feel this is an important edge case to handle. This could happen for any split operation where the shard to split has non-null parent shards and only one of its children has a lease. This could be fairly common as most shards in a frequently resharded stream have non-null parents.